core/store/models/service_agreement.go

```diff
@@ -73,109 +104,84 @@ func BuildServiceAgreement(us UnsignedServiceAgreement, signer Signer) (ServiceA
-// sign a service agreement
-func NewUnsignedServiceAgreementFromRequest(reader io.Reader) (UnsignedServiceAgreement, error) {
-	var jsr JobSpecRequest
-
+func parseServiceAgreementJSON(reader io.Reader) (
+	string, ServiceAgreementRequest, error) {
 	input, err := ioutil.ReadAll(reader)
 	if err != nil {
-		return UnsignedServiceAgreement{}, errors.Wrap(err, "while reading service agreement JSON")
+		return "", ServiceAgreementRequest{}, errors.Wrap(err,
+			"while reading service agreement JSON")
 	}
-
-	err = json.Unmarshal(input, &jsr)
+	var sar ServiceAgreementRequest
+	err = json.Unmarshal(input, &sar)
 	if err != nil {
-		return UnsignedServiceAgreement{}, errors.Wrap(err, "while parsing job-spec JSON")
-	}
-
-	var encumbrance Encumbrance
-	if err := json.Unmarshal(input, &encumbrance); err != nil {
-		return UnsignedServiceAgreement{}, errors.Wrap(err, "while parsing service agreement JSON")
-	}
-
-	if encumbrance.Aggregator == "" {
-		return UnsignedServiceAgreement{}, fmt.Errorf("must set aggregator contract address")
+		return "", ServiceAgreementRequest{}, errors.Wrap(err,
+			"while parsing service agreement JSON")
 	}
-
 	normalized, err := utils.NormalizedJSON(input)
 	if err != nil {
-		return UnsignedServiceAgreement{}, errors.Wrap(err, "while normalizing service agreement JSON")
+		return "", ServiceAgreementRequest{}, errors.Wrap(err,
+			"while normalizing service agreement JSON")
 	}
+	return normalized, sar, nil
+}
 
-	requestDigest, err := utils.Keccak256([]byte(normalized))
+// NewUnsignedServiceAgreementFromRequest builds the information required to
+// sign a service agreement
+func NewUnsignedServiceAgreementFromRequest(reader io.Reader) (UnsignedServiceAgreement, error) {
+	normalized, sar, err := parseServiceAgreementJSON(reader)
 	if err != nil {
 		return UnsignedServiceAgreement{}, err
 	}
+	us := UnsignedServiceAgreement{
+		Encumbrance: Encumbrance{
+			Payment:                sar.Payment,
+			Expiration:             sar.Expiration,
+			EndAt:                  sar.EndAt,
+			Oracles:                sar.Oracles,
+			Aggregator:             sar.Aggregator,
+			AggInitiateJobSelector: sar.AggInitiateJobSelector,
+			AggFulfillSelector:     sar.AggFulfillSelector,
+		},
+		RequestBody: normalized,
+		JobSpecRequest: JobSpecRequest{
+			Initiators: sar.Initiators,
+			Tasks:      sar.Tasks,
+			StartAt:    null.NewTime(sar.StartAt.Time, sar.StartAt.Valid),
+			EndAt:      null.NewTime(sar.EndAt.Time, sar.EndAt.Valid),
+			MinPayment: sar.Payment,
+		},
+	}
 
-	id, err := generateServiceAgreementID(encumbrance, common.BytesToHash(requestDigest))
+	requestDigest, err := utils.Keccak256([]byte(normalized))
 	if err != nil {
 		return UnsignedServiceAgreement{}, err
 	}
 
-	us := UnsignedServiceAgreement{
-		ID:             id,
-		Encumbrance:    encumbrance,
-		RequestBody:    normalized,
-		JobSpecRequest: jsr,
+	us.ID, err = generateServiceAgreementID(us.Encumbrance,
```

Comment:
In JS/SOL land, we use the `SAID` acronym instead of the expanded name. Should we do the same here or change the other files to use the full name? Personally I like the more verbose naming.

Reply:
In general I'm in favor of more verbose naming, but it can get cumbersome. I think `ServiceAgreementID` everywhere would be one such case, but I don't feel too strongly about it if you want to change it.

Reply:
Ill rename this to `generateSAID` then, I care about consistency more than anything else TBH